### ISSUE TITLE
Perform aggregation in the database to speed up Whiterabbit

### ISSUE THREAD

**Issue by ablack3** [OHDSI/WhiteRabbit#325]
Forgive me if this idea has already been considered or if I am misunderstanding how whiterabbit works. 
Whiterabbit takes a long time to run and I think one reason is that it downloads a lot of data and then processes (e.g. counts unique values) in the downloaded sample. Would it be possible to let the database handle the processing and then only download the result so that we could take advantage of in-database optimizations (e.g. fast counting unique values in a single column)?

So instead of 
`query = "SELECT * FROM " + table + " SAMPLE(" + percentage + ")";`
and then processing the values of each column in java whiterabbit scan would use a set of queries like
`select {fieldname}, count({fieldname}) from {table}`

This isn't a well-formed alternative but it is meant to communicate the idea of doing the aggregation in the database rather than downloading a large sample of the data as a way to speed up the database scan. Calculations like MIN, MAX, and MEAN would be handled in the database as well.

**Comment by blootsvoets** [comment]
Performance-wise, this is a great idea. Practically, this can only be done by limiting the number of database types that White Rabbit can read, because we don’t have the capacity to test the queries in each SQL dialect. The current implementation would still be needed for CSV input data. The code would need some refactoring to allow for the different aggregation methods.

**Comment by ablack3** [comment]
> Practically, this can only be done for a limited number of database types that White Rabbit, because we don’t have the capacity to test the queries in each SQL dialect.

The [SqlRender package](https://github.com/OHDSI/SqlRender) has a java interface and can be used to translate the aggregation queries to any OHDSI supported dialect. Current supported dialects are Microsoft SQL Server, Oracle, PostgreSQL, Microsoft Parallel Data Warehouse, Apache Impala, IBM Netezza, Google BigQuery, Apache Spark, SQLite, and Amazon RedShift.

> The current implementation would still be needed for CSV input data.

What is the CSV input data? Is this for the ETL unit test functionality? If so I think it would still be worth speeding up the scan report since the scan report and ETL unit tests can be used independently.

**Comment by ablack3** [comment]
In a recent test on redshift doing in database aggregation is very fast for a single column with billions of rows. Downloading a sample of data using `ORDER BY RANDOM()` is too slow to be feasible. WhiteRabbit should take advantage of column oriented database optimizations when possible.

**Comment by MaximMoinat** [comment]
Agreed, this would be a nice enhancement. It does require a considerable amount of refactoring and thorough testing. We should first assemble a small team with access to all database systems supported by WR. @ablack3 Could you take the lead to form such a team?

> What is the CSV input data? Is this for the ETL unit test functionality? If so I think it would still be worth speeding up the scan report since the scan report and ETL unit tests can be used independently.
CSV is simply one of the formats a dataset can come in. Either exported from a database or from Excel (yes, some databases like registries are kept entirely in Excel).

**Comment by ablack3** [comment]
I don't know about leading a team but I would like to contribute to the solution. Ideally I'd like to do some of the programming with supervision from someone who has worked on this application for a while.
I propose starting with a team of two: me and @MaximMoinat or me and @blootsvoets. 
We start with an orientation to the WhiteRabbit code base since I am very new the java side of OHDSI software.
We then outline the changes which would include addition of a column oriented scan and automated testing on each database platform using github actions and the databases that @leeevans has set up. To my knowledge no individual or organization has access to all supported platforms which is unfortunate. 

Once we scope the changes and estimate the work required then we revisit the question of actually making the change and maybe separate the testing and "in database aggregation" into two separate steps.

**Comment by ablack3** [comment]
Pointer to the relevant code https://github.com/OHDSI/WhiteRabbit/blob/062891422b047baccb174bd9c47fa0984670aa84/whiterabbit/src/main/java/org/ohdsi/whiteRabbit/scan/SourceDataScan.java#L427

tagging @mgabetta

**Comment by mgabetta** [comment]
Hi @ablack3 and All,

recently I've supported Prof. Cristiana Larizza form University of Pavia (@cristianalarizza) to help her moving some operation, currently performed by WhiteRabbit in memory, to the source database side, thus reducing the workload on the Java application.

Cristiana's work focused on:
- separating the execution of analysis tasks on column basis (instead of making it table per table)
- executing aggregation on numeric fields (avg, median, sd, quartiles...) on the database
- running all sampling tasks (that generate table specific Excel sheets) with GROUP BYs and COUNTs on the the database side.

The work is still in progress, but preliminary results are promising: on a small Synthea DB the improved version performs scan 4 times faster than standard WR.

Next steps will involve:
- extension of the new logic to more DBMS (we've developed only for PostgreSQL at the moment)
- extensive testing with all possible input params from the GUI (at the moment some of them have been overlooked)
- code refactoring
- test on a larger databases (also for Postgres)

Regarding the last point: do you have any suggestion about publicly available database than can serve the purpose?

**Comment by ablack3** [comment]
Awesome! I'm not aware of any publicly available large CDM databases. I do know that if you set up OHDSIonAWS you can select large synpuf and sythea OMOP CDMs to be created in Redshift. But this requires an AWS account and has an associated cost. It is not too expensive to spin up, run tests, and shut down but you don't want to leave it on very long! https://github.com/OHDSI/OHDSIonAWS
